### DISCUSSION TITLE
[Question] Questions on Theory of Arrays

### DISCUSSION THREAD

**Discussion by Lee-Janggun** [prosyslab-classroom/cs424-program-reasoning#368] [Q&A]
- In class on 9/29/2022, we learned about the theory of arrays and the key axioms.
- One thing I noticed but forgot to mention is that there does not seem to be any check that indexes are inbounds. Does this mean that the theory is for unbounded arrays?
- In the case that we add a restriction for bounded arrays, would we need to be careful to not assert anything about out-of-bounds access, or will it be OK to say it is None?
- Another impression I got was that we implicitly assumed array was "initialized". Would a theory where a read on a index was only possible after a write was done make sense?

**Comment by KihongHeo** [comment]
Hi.
Interesting questions. Also, other students asked the same question after the class.

In the theory, there is no concept of `length`. However, it does not mean we only consider "unbounded" arrays. It means we cannot reason about anything related to `length` (so, also about `out-of-bound`). They are unknown.  Similarly, for `initialization`, we cannot reason about anything related to `initialization`. Notice that the axioms do not describe anything about different arrays but the same array (`a`). So, the length is not our interest here.

You may be able to come up with more powerful theories something like `the Theory of C Arrays`. But I believe the theory we discussed today is enough to reason about many properties of programs.

**Comment by Lee-Janggun** [comment]
Thank you for the detailed answer!
